File: Backtracking/90_Subsets_II/90_Subsets_II.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> current = new ArrayList<>();
        backtrack(0, current, res, nums);
        return res;
    }
    
    private void backtrack(int start, List<Integer> current, List<List<Integer>> res, int[] nums) {
        res.add(new ArrayList<>(current));
        for(int i = start; i < nums.length; i++) {
            if(i > start && nums[i] == nums[i - 1]) {
                continue;
            }
            current.add(nums[i]);
            backtrack(i + 1, current, res, nums);
            current.remove(current.size() - 1);
        }
    }
}

File: Backtracking/90_Subsets_II/90_Subsets_II.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<int> current;
        vector<vector<int>> res;
        backtrack(0, current, res, nums);
        return res;
    }
private: 
    void backtrack(size_t start, vector<int>& current, vector<vector<int>>& res, vector<int>& nums) {
        res.push_back(current);
        for( size_t i=start; i<nums.size(); i++) {
            if(i>start && nums[i] == nums[i-1])
                continue;
            current.push_back(nums[i]);
            backtrack(i+1, current, res, nums);
            current.pop_back();
        }
    }
};
